packages/txservice/src/provider.ts

```diff
@@ -0,0 +1,177 @@
+import { ChainError } from "./error";
+import { FullTransaction, MinimalTransaction, ProviderConfig } from "./types";
+
+const { JsonRpcProvider, FallbackProvider } = providers;
+
+/// Could use a more encompassing name, e.g. ChainRpcDispatch, etc
+export class ChainRpcProvider {
+  private provider: providers.JsonRpcProvider | providers.FallbackProvider;
+  private config: { gasInitialBumpPercent: number; gasPriceMinimum: BigNumber; rpcProviderMaxRetries: number; };
+  private signer: Signer;
+  private queue: PriorityQueue = new PriorityQueue({ concurrency: 1 });
+
+  public confirmationsRequired: number;
+  public chainId: number;
+
+  constructor(
+    private readonly log: BaseLogger,
+    chainId: number,
+    signer: string | Signer,
+    providerConfigs: ProviderConfig[],
+    config: TransactionServiceConfig,
+  ) {
+    const { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries } = config;
+    this.config = { gasInitialBumpPercent, gasPriceMinimum, rpcProviderMaxRetries };
+    this.confirmationsRequired = config.chainConfirmations.get(chainId) ?? config.defaultConfirmationsRequired;
+    this.chainId = chainId;
+    
+    // Register a provider for each url.
+    // Make sure all providers are ready()
+
+    if (providerConfigs.length === 1) {
+      this.provider = new JsonRpcProvider(providerConfigs[0]);
+    } else {
+      const fallbackProviderConfigs = providerConfigs.map((config) => ({
+        provider: new JsonRpcProvider(
+          {
+            url: config.url,
+            user: config.user,
+            password: config.password,
+          }
+        ),
+        priority: config.priority ?? 1,
+        weight: config.weight ?? 1,
+        stallTimeout: config.stallTimeout,
+      }));
+      // TODO: Quorum is set to 1 here, but we may want to reconfigure later. Normally it is half the sum of the weights,
+      // which might be okay in our case, but for now we have a low bar.
+      this.provider = new FallbackProvider(fallbackProviderConfigs, 1);
+    }
+
+    this.signer = typeof signer === "string" ? new Wallet(signer, this.provider) : signer.connect(this.provider);
+  }
+
+  public async sendTransaction(
+    tx: FullTransaction
+  ): Promise<{ response: providers.TransactionResponse | Error; success: boolean }> {
+    const ready = await this.provider.ready;
+    if (!ready) {
+      // Error out, no providers are ready.
+      throw new ChainError(ChainError.reasons.ProviderNotReady);
+    }
+    // TODO: Check getBlockNumber() > 0.
+
+    // Define task to send tx with proper nonce
+    const task = async (): Promise<{ response: providers.TransactionResponse | Error; success: boolean }> => {
+      try {
+        // Send transaction using the passed in callback.
+        // const stored = this.nonces.get(chainId);
+        // const nonceToUse: number = nonce ?? stored ?? (await signer.getTransactionCount("pending"));
+        const pending: number = await this.signer.getTransactionCount("pending");
```

Comment:
Yea. I'm thinking that we can achieve a cleaner approach here than what we had before since we are using the `Transaction` class, which can manage its own nonce once it's assigned one.

That being said, we still run into the problem of `this.signer.getTransactionCount("pending");` not returning the correct nonce or something. I also have to look over the latest code on main bc I know layne changed some nonce-related things (I vaguely recall from last week, but need refresher)

Reply:
Okay, see new solution. We are using `NonceManager` now, which should resolve our issues in getting nonce as it increments it internally. Worth some testing, ofc